cmd/kubeadm/app/util/apiclient/wait.go

```diff
@@ -235,12 +233,9 @@ func getStaticPodSingleHash(client clientset.Interface, nodeName string, compone
 		return "", err
 	}
 
-	podBytes, err := json.Marshal(staticPod)
-	if err != nil {
-		return "", err
-	}
-
-	return fmt.Sprintf("%x", sha256.Sum256(podBytes)), nil
+	staticPodHash := staticPod.Annotations["kubernetes.io/config.hash"]
```

Comment:
I found this very interesting.
The `config.hash` annotation equals the UID of the kubelet static pod, but the mirror-pod has its own UID:
```shell
# mirror pod from the apiserver
root@vagrant:~# kubectl -n kube-system get po etcd-vagrant -ojson | jq -S .metadata
{
  "annotations": {
    "kubernetes.io/config.hash": "7278f85057e8bf5cb81c9f96d3b25320",
    "kubernetes.io/config.mirror": "7278f85057e8bf5cb81c9f96d3b25320",
    "kubernetes.io/config.seen": "2018-04-01T03:06:29.912626062Z",
    "kubernetes.io/config.source": "file",
    "scheduler.alpha.kubernetes.io/critical-pod": ""
  },
  "creationTimestamp": "2018-04-01T03:07:53Z",
  "labels": {
    "component": "etcd",
    "tier": "control-plane"
  },
  "name": "etcd-vagrant",
  "namespace": "kube-system",
  "resourceVersion": "33491",
  "selfLink": "/api/v1/namespaces/kube-system/pods/etcd-vagrant",
  "uid": "deb48bce-3559-11e8-91d2-080027d6be16"
}

# podSpec from the kubelet API
root@vagrant:~# curl -s 127.0.0.1:10255/pods | jq -S '.items[]|select(.metadata.name|contains("etcd")) | .metadata'
{
  "annotations": {
    "kubernetes.io/config.hash": "7278f85057e8bf5cb81c9f96d3b25320",
    "kubernetes.io/config.seen": "2018-04-01T03:10:32.494811778Z",
    "kubernetes.io/config.source": "file",
    "scheduler.alpha.kubernetes.io/critical-pod": ""
  },
  "creationTimestamp": null,
  "labels": {
    "component": "etcd",
    "tier": "control-plane"
  },
  "name": "etcd-vagrant",
  "namespace": "kube-system",
  "selfLink": "/api/v1/namespaces/kube-system/pods/etcd-vagrant",
  "uid": "7278f85057e8bf5cb81c9f96d3b25320"
}
```

Regular pods don't seem to show up in the kubelet API `/pods` and do not have the `config.hash` annotation:
```
root@vagrant:~# kubectl run alpine --image=alpine --restart=Never --rm -- sleep 1000
error: --rm should only be used for attached containers
See 'kubectl run -h' for help and examples.
root@vagrant:~# kubectl run alpine --image=alpine --restart=Never -- sleep 1000
pod "alpine" created
root@vagrant:~# k get po alpine -ojson | jq -S .metadata
{
  "creationTimestamp": "2018-04-03T00:35:44Z",
  "labels": {
    "run": "alpine"
  },
  "name": "alpine",
  "namespace": "default",
  "resourceVersion": "34635",
  "selfLink": "/api/v1/namespaces/default/pods/alpine",
  "uid": "f22d81c8-36d6-11e8-8286-080027d6be16"
}
root@vagrant:~# curl -s 127.0.0.1:10255/pods | jq -S '.items[]|select(.metadata.name|contains("alpine")) | .metadata'
root@vagrant:~# curl -s 127.0.0.1:10255/pods | grep alpine
root@vagrant:~#
```

Reply:
> The config.hash annotation equals the UID of the kubelet static pod

@stealthybox this actually makes sense because the default pod UID is generated with`hash.DeepHashObject`: https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/kubelet/config/common.go#L60

> Regular pods don't seem to show up in the kubelet API /pods and do not have the config.hash annotation

Interesting...  I hadn't gotten this far with my testing last week (and will not have time to do any additional digging this week). There are additional endpoints that the kubelet exposes that might provide the information we need.

Reply:
This is expected but I'm not sure about the reason/details. @yujuhong could you please explain a bit about this? Thanks!

Reply:
I think there's two viable keys for determining this change:
- `.metadata.annotations.kubernetes.io/config.hash`
  equal between mirror-pod and kubelet static podSpec
- `.metadata.uid`
  different between mirror-pod and kubelet static podSpec

I believe both of these fields will result in similar behavior since a new static-pod UID should also mean a new mirror-pod UID, but I haven't verified this.

I'm +1 on the current implementation which uses the config.hash annotation since it's uniform between both data-structures.